src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectWorkspaceStateGenerator.TestAccessor.cs

```diff
@@ -0,0 +1,62 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+
+namespace Microsoft.VisualStudio.Razor;
+
+internal sealed partial class ProjectWorkspaceStateGenerator
+{
+    internal TestAccessor GetTestAccessor() => new(this);
+
+    internal sealed class TestAccessor(ProjectWorkspaceStateGenerator instance)
+    {
+        public interface IUpdateItem
+        {
+            bool IsCompleted { get; }
+            bool IsCancellationRequested { get; }
+        }
+
+        private sealed class UpdateItemWrapper(UpdateItem updateItem) : IUpdateItem
+        {
+            public bool IsCompleted => updateItem.UpdateTask.IsCompleted;
+            public bool IsCancellationRequested => updateItem.IsCancellationRequested;
+        }
+
+        public ImmutableArray<IUpdateItem> GetUpdates()
+        {
+            lock (instance._updates)
+            {
+                using var result = new PooledArrayBuilder<IUpdateItem>(capacity: instance._updates.Count);
```

Comment:
Your test-only code is more memory efficient than any single piece of product code I wrote from 2004 to 2018 😁